mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/task/RecomputeCurrentQuotasService.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.james.util.streams.Iterators;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+public class RecomputeCurrentQuotasService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RecomputeCurrentQuotasService.class);
+
+    static class Context {
+        static class Snapshot {
+            private final long processedQuotaRootCount;
+            private final ImmutableList<QuotaRoot> failedQuotaRoots;
+
+            private Snapshot(long processedQuotaRootCount, ImmutableList<QuotaRoot> failedQuotaRoots) {
+                this.processedQuotaRootCount = processedQuotaRootCount;
+                this.failedQuotaRoots = failedQuotaRoots;
+            }
+
+            long getProcessedQuotaRootCount() {
+                return processedQuotaRootCount;
+            }
+
+            ImmutableList<QuotaRoot> getFailedQuotaRoots() {
+                return failedQuotaRoots;
+            }
+        }
+
+        private final AtomicLong processedQuotaRootCount;
+        private final ConcurrentLinkedDeque<QuotaRoot> failedQuotaRoots;
+
+        Context() {
+            processedQuotaRootCount = new AtomicLong();
+            failedQuotaRoots = new ConcurrentLinkedDeque<>();
+        }
+
+        void incrementProcessed() {
+            processedQuotaRootCount.incrementAndGet();
+        }
+
+        void addToFailedMailboxes(QuotaRoot quotaRoot) {
+            failedQuotaRoots.add(quotaRoot);
+        }
+
+        Snapshot snapshot() {
+            return new Snapshot(processedQuotaRootCount.get(),
+                ImmutableList.copyOf(failedQuotaRoots));
+        }
+    }
+
+    private final UsersRepository usersRepository;
+    private final StoreCurrentQuotaManager storeCurrentQuotaManager;
+    private final CurrentQuotaCalculator currentQuotaCalculator;
+    private final UserQuotaRootResolver userQuotaRootResolver;
+    private final SessionProvider sessionProvider;
+
+    @Inject
+    public RecomputeCurrentQuotasService(UsersRepository usersRepository,
+                                         StoreCurrentQuotaManager storeCurrentQuotaManager,
+                                         CurrentQuotaCalculator currentQuotaCalculator,
+                                         UserQuotaRootResolver userQuotaRootResolver,
+                                         SessionProvider sessionProvider) {
+        this.usersRepository = usersRepository;
+        this.storeCurrentQuotaManager = storeCurrentQuotaManager;
+        this.currentQuotaCalculator = currentQuotaCalculator;
+        this.userQuotaRootResolver = userQuotaRootResolver;
+        this.sessionProvider = sessionProvider;
+    }
+
+    Task.Result recomputeCurrentQuotas(Context context) {
+        try {
+            return Iterators.toStream(usersRepository.list())
+                .map(username -> recomputeUserCurrentQuotas(context, username))
```

Comment:
Maybe we could have a `QuotaRootResolver::listQuotaRoots` method?

That would avoid hardcoding the user <-> quotaRoot relation here?

Reply:
Yes your remark makes sense, thanks

Reply:
Checked a bit, seems making it more complex for couple of test classes if I add the UsersRepository in QuotaRootResolver for listQuotaRoots... I will see at the end